### PR TITLE
Keep tool parameters when rerunning a job

### PR DESCRIPTION
**Pull request by mvdbeek**
with a different tool version, as long as the parameter did not change.
Addresses https://github.com/galaxyproject/galaxy/issues/1729

### CODE REVIEW

File: client/galaxy/scripts/mvc/tool/tool-form-base.js [hunk 1]
```diff
@@ -66,14 +66,14 @@ define( [ 'utils/utils', 'utils/deferred', 'mvc/ui/ui-misc', 'mvc/form/form-view
...
+                build_url = Galaxy.root + 'api/jobs/' + options.job_id + '/build_for_rerun';
             } else {
                 build_url = Galaxy.root + 'api/tools/' + options.id + '/build';
-                build_data = $.extend( {}, Galaxy.params );
-                build_data[ 'tool_id' ] && ( delete build_data[ 'tool_id' ] );
-                options.version && ( build_data[ 'tool_version' ] = options.version );
             }
+           build_data = $.extend( {}, Galaxy.params );
+           build_data[ 'tool_id' ] && ( delete build_data[ 'tool_id' ] );
+           options.version && ( build_data[ 'tool_version' ] = options.version );
```
Comment (guerler) [inline client/galaxy/scripts/mvc/tool/tool-form-base.js]: Changes to this file look like a bad merge.

Comment (mvdbeek) [reply client/galaxy/scripts/mvc/tool/tool-form-base.js]: @guerler I've rebased the PR (36bdd99), does this look better to you ?

File: client/galaxy/scripts/mvc/tool/tool-form-base.js [hunk 2]
```diff
@@ -66,14 +66,14 @@ define( [ 'utils/utils', 'utils/deferred', 'mvc/ui/ui-misc', 'mvc/form/form-view
...
-                build_url = Galaxy.root + 'api/jobs/' + new_options.job_id + '/build_for_rerun';
+            if ( options.job_id ) {
+                build_url = Galaxy.root + 'api/jobs/' + options.job_id + '/build_for_rerun';
             } else {
                 build_url = Galaxy.root + 'api/tools/' + options.id + '/build';
                 build_data = $.extend( {}, Galaxy.params );
                 build_data[ 'tool_id' ] && ( delete build_data[ 'tool_id' ] );
-                options.version && ( build_data[ 'tool_version' ] = options.version );
             }
+           options.version && ( build_data[ 'tool_version' ] = options.version );
```
Comment (nsoranzo) [inline client/galaxy/scripts/mvc/tool/tool-form-base.js]: Seems the indentation is missing a space.

File: client/galaxy/scripts/mvc/tool/tool-form-base.js [hunk 3]
```diff
@@ -66,14 +66,14 @@ define( [ 'utils/utils', 'utils/deferred', 'mvc/ui/ui-misc', 'mvc/form/form-view
             // build request url
             var build_url = '';
             var build_data = {};
-            if ( new_options.job_id ) {
-                build_url = Galaxy.root + 'api/jobs/' + new_options.job_id + '/build_for_rerun';
+            if ( options.job_id ) {
+                build_url = Galaxy.root + 'api/jobs/' + options.job_id + '/build_for_rerun';
```
Comment (nsoranzo) [inline client/galaxy/scripts/mvc/tool/tool-form-base.js]: These 2 changes above seem to revert the recent commit https://github.com/galaxyproject/galaxy/commit/a498390bef74301816945551a9832f38747fd4fc from @guerler which I think was correct, still a bad merge maybe?

File: lib/galaxy/tools/toolbox/base.py [hunk 4]
```diff
@@ -418,7 +418,8 @@ def get_tool( self, tool_id, tool_version=None, get_all_versions=False, exact=Fa
                 if tool_version and tool_version in self._tool_versions_by_id[ tool_id ]:
                     return self._tool_versions_by_id[ tool_id ][ tool_version ]
                 # tool_id exactly matches an available tool by id (which is 'old' tool_id or guid)
-                return self._tools_by_id[ tool_id ]
+                if not tool_version:
+                    return self._tools_by_id[ tool_id ]
```
Comment (nsoranzo) [inline lib/galaxy/tools/toolbox/base.py]: Good catch! I'd restructure lines 418-422 as:
```python
                # tool_id exactly matches an available tool by id (which is 'old' tool_id or guid)
                if not tool_version:
                    return self._tools_by_id[ tool_id ]
                elif tool_version in self._tool_versions_by_id[ tool_id ]:
                    return self._tool_versions_by_id[ tool_id ][ tool_version ]
```

File: client/galaxy/scripts/mvc/tool/tool-form-base.js (hunk 3 above)
Comment (mvdbeek) [reply client/galaxy/scripts/mvc/tool/tool-form-base.js]: What I have observed is that `options` always contains the `job_id`, but `new_options` only contains changed options. `new_options` only contains the `job_id` when clicking on the rerun icon. So the change in a498390 narrows looking up the `job_id`, but we now also need the `job_id` when having clicked on the rerun icon, followed by switching the tool version (It is not a new option anymore). So maybe it is better to use either `new_options.job_id` or `options.job_id`

**Review by nsoranzo — APPROVED** [review]

**Comment by jmchilton** [comment]
Great - as always thanks a ton @mvdbeek !

**Pull request merged**